In [ ]:
#calculate PV and FV
def calculate_future_value(PV, rate, periods):
    return PV * (1 + rate)**periods

def calculate_present_value(FV, rate, periods):
    return FV / (1 + rate)**periods
initial_investment=
initial_rate=
maturity=
future_value = calculate_future_value(initial_investment, initial_rate, maturity)
present_value = calculate_present_value(future_value, initial_rate, maturity)
print(f"Future value of the securities: ${future_value:,.2f}")
print(f"Present value of the securities after the interest rate change: ${new_present_value:,.2f}")

In [ ]:
#calculate NPV
def calculate_npv(cash_flow, interest_rate):
    npv = 0
    for i, score in enumerate(cash_flow):
        npv += score / (1 + interest_rate) ** i
    return npv

# Define the happiness scores for each career path
career = [100 + 10 * i for i in range(5)]#Or something else, plug in the cash flow stream
discount_rate = 0.05
npv = calculate_npv(career, discount_rate)
print(f"Career Path A happiness scores: {career}")

In [2]:
#calculate IRR
import numpy as np
from scipy.optimize import fsolve


def IRRPoly(cash_flow):
    # Create the polynomial coefficients
    coefficients = np.flip(cash_flow)
    
    # Find the roots of the polynomial
    roots = np.roots(coefficients)
    
    # Find the real positive root
    irr_poly = np.real(roots[(roots >= 0)  & (np.iscomplex(roots) == False)])[0] 
    
    # Solve for C=1/(1+r)
    irr= (1-irr_poly)/irr_poly

    return irr

def IRR(cash_flow):
    # Define the NPV function
    def npv_function(r, cash_flow):
        n = len(cash_flow)
        return sum(cash_flow[i] / (1 + r)**i for i in range(n))

    # Define the objective function for the IRR
    def objective_function(r, cash_flow):
        return npv_function(r, cash_flow)

    result = fsolve(objective_function, 0.1, args=(cash_flow))
    irr = result[0]
    return(irr)
#I prefer to use the second method, the first one only works for yearly IRR
#cash flow=[...]
#IRR1=IRRPoly(cash_flow)
a=IRRPoly([-50,20,20,20,20,20])
b=IRR([-50,20,20,20,20,20])
print(a,b)

# Print out solution
#print(f"The investment has an internal rate of return of {IRR1 * 100:.0f}%.")

0.2864929024976756 0.2864929024976759


In [4]:
# spot rate, forward rate
import numpy as np

# What we are given
S1 = 0.023
S2 = 0.029
f45=0.1155
S4=0.069
#S5=0.069
#S2=np.sqrt((1+S3)**3/(1+f23))-1
f12 = (1+S2)**2/(1+S1)-1
ftest=((1+S4)**2*((1+f45))**2)**(1/4)-1
#f45 = (1+S5)**5/(1+S4)**4-1
print(ftest)

#print(f"S2: {S2:.4f}")
print(f"f12: {f12:.4f}")
#print(f"S5: {S5:.4f}")
#print(f"f45: {f45:.4f}")
#s1,s2->f12 (1+S2)**2/(1+S1)-1
#s4,s5->f45  (1+S5)**5/(1+S4)**4-1
#s4,f45->s5  ((1+S4)**4*((1+f45)))**(1/5)-1
#f45,s5->s4  ((1+S5)**5/(1+f45))**(1/4)-1

0.09200251831211448
f12: 0.0350


In [ ]:
#calculate bond price and Macaulay duration
import numpy as np

def bond_price(F, C, lamb, m, n):
    '''
    F is the face value of the bond
    C is the yearly coupon payment
    lamb is the yield of the bond 
    m is the number of coupon payments per year
    n is the number of coupon payments remaning
    '''
    discount_factor = (1 + lamb / m) ** n
    bond_price = F / discount_factor + (C / lamb) * (1 - 1 / discount_factor)
    return bond_price
def macaulay_duration(c, y, m, n):
    """
    c is the coupon rate per period
    y is the yield per period
    m is the number of periods per year
    n is the number of periods remaining
    """
    
    D = ((1+y)/(m*y)) - (1+y+n*(c-y))/(m*c*((1+y)**n-1)+m*y)
    return D

# Bond information
face_value = 1000
yield_per_annum = 0.05
coupon_rate_A = 0.04
maturity_A = 10
payment_frequency = 2

# Calculate the initial investment required for the bond portfolio
price_A = bond_price(face_value, face_value * coupon_rate_A , yield_per_annum, payment_frequency, maturity_A * payment_frequency)

# Calculate the Macaulay duration for both bonds
duration_A = macaulay_duration(coupon_rate_A / payment_frequency, yield_per_annum/ payment_frequency, payment_frequency, maturity_A * payment_frequency)

In [3]:
# calculate monthly payment and balance
#换了两年后利率降低了 dis3-1

# Step 1: Calculate the original monthly payment
loan_amount = 30000
original_interest_rate = 0.06 / 12
original_loan_period = 10 * 12

def calculate_monthly_payment(P, r, n):
    return P * r * (1 + r)**n / ((1 + r)**n - 1)

def calculate_remaining_balance(A, r, n):
    remaining_balance = (A/r)*(1-1/(1+r)**n)
    return remaining_balance


original_monthly_payment = calculate_monthly_payment(loan_amount, original_interest_rate, original_loan_period)
print(f'Original monthly payment: ${original_monthly_payment:.2f}')

# Step 2: Calculate the remaining balance after 2 years
payments_to_go = original_loan_period-(2 * 12)


remaining_balance = calculate_remaining_balance(original_monthly_payment, original_interest_rate, payments_to_go)
print(f'Remaining balance after 2 years: ${remaining_balance:.2f}')

# Step 3: Calculate the new monthly payment after refinancing
new_interest_rate = 0.045 / 12
new_loan_period = 15 * 12

new_monthly_payment = calculate_monthly_payment(remaining_balance, new_interest_rate, new_loan_period)
print(f'New monthly payment: ${new_monthly_payment:.2f}')

def calculate_npv(cash_flow, r):
    npv = 0
    for i, score in enumerate(cash_flow):
        npv += score / (1 + r) ** i
    return npv

Original monthly payment: $333.06
Remaining balance after 2 years: $25344.39
New monthly payment: $193.88
Net Present Value for not refinancing: $28466.01
Net Present Value for refinancing: $28145.49


In [4]:
#calculate bond's YTM
import numpy as np

def IRRPoly(cash_flow, m):
    coefficients = np.flip(cash_flow)
    roots = np.roots(coefficients)
    irr_poly = np.real(roots[(roots >= 0)  & (np.iscomplex(roots) == False)])[0]
    irr = m*(1 - irr_poly) / irr_poly
    return irr

# Bond characteristics
face_value = 10000
coupon_payment = 500#这里是payment，不是coupon rate
payments_per_year = 2
years_to_maturity = 10
current_price = 10800 #P current market price

# Cash flows for the bond

cash_flows = (coupon_payment / payments_per_year)*np.ones(years_to_maturity * payments_per_year)
cash_flows[-1] += face_value
print(cash_flows)
cash_flows = np.insert(cash_flows, 0, -current_price)
print(cash_flows)

# Find YTM using IRRPoly function
ytm = IRRPoly( cash_flows, payments_per_year)

# Convert to a percentage and display the result
ytm_percentage = ytm * 100
print(f"The bond's yield to maturity is {ytm_percentage:.2f}%")

[  250.   250.   250.   250.   250.   250.   250.   250.   250.   250.
   250.   250.   250.   250.   250.   250.   250.   250.   250. 10250.]
[-10800.    250.    250.    250.    250.    250.    250.    250.    250.
    250.    250.    250.    250.    250.    250.    250.    250.    250.
    250.    250.  10250.]
The bond's yield to maturity is 4.02%


In [8]:
#calculate bond price
import numpy as np
import matplotlib.pyplot as plt

def bond_price(F, C, ytm, m, n):
    '''
    F is the face value of the bond
    C is the yearly coupon payment
    lamb is the yield of the bond 
    m is the number of coupon payments per year
    n is the number of coupon payments remaning
    '''
    discount_factor = (1 + ytm / m) ** n
    bond_price = F / discount_factor + (C / ytm) * (1 - 1 / discount_factor)
    return bond_price

face_value = 1000
coupon_rate = 0.02
years_to_maturity = 10
frequency = 2
ytm = 0.0402
#np.linspace(0.01, 0.15, 100)
coupon_payment_per_period = face_value * coupon_rate / frequency
price = bond_price(face_value, coupon_payment_per_period, ytm, frequency, years_to_maturity * frequency)
print("price=",price)

price= 753.3314625269959


In [ ]:
#calculate bond price,D,MD, estimated new bond price and the comparsion
def bond_price(F, C, ytm, m, n):
    '''
    F is the face value of the bond
    C is the yearly coupon payment
    lamb is the yield of the bond 
    m is the number of coupon payments per year
    n is the number of coupon payments remaning
    '''
    discount_factor = (1 + ytm / m) ** n
    bond_price = F / discount_factor + (C / ytm) * (1 - 1 / discount_factor)
    return bond_price
def macaulay_duration(c, y, m, n):
    """
    c is the coupon rate per period
    y is the yield per period
    m is the number of periods per year
    n is the number of periods remaining
    """
    
    D = ((1+y)/(m*y)) - (1+y+n*(c-y))/(m*c*((1+y)**n-1)+m*y)
    return D
face_value = 1000
coupon_rate = 0.06
maturity_years = 8
frequency = 2
current_ytm = 0.04
new_ytm = 0.05

coupon_payment_per_year = face_value * coupon_rate 
coupon_rate_per_period=  coupon_rate /frequency
yield_to_maturity_per_period = current_ytm / frequency
periods = maturity_years * frequency

# Step 1: Calculate the bond price
bondPrice = bond_price(face_value, coupon_payment_per_year, current_ytm, frequency, periods)
print(f"1. Bond Price: {bondPrice:.2f}")

# Step 2: Calculate the Macaulay duration
macaulayDuration = macaulay_duration(coupon_rate_per_period, yield_to_maturity_per_period, frequency, periods)
print(f"2. Macaulay Duration: {macaulayDuration:.2f}")

# Step 3: Calculate the modified duration
modifiedDuration = macaulayDuration / (1 + yield_to_maturity_per_period)
print(f"3. Modified Duration: {modifiedDuration:.2f}")

# Step 4: Estimate the new bond price using the modified duration
delta_ytm = new_ytm - current_ytm
estimated_new_bond_price = bondPrice * (1 - modifiedDuration * delta_ytm)
print(f"4. Estimated New Bond Price: {estimated_new_bond_price:.2f}")

# Step 5: Calculate the actual new bond price using the updated YTM
actual_new_bond_price = bond_price(face_value, coupon_payment_per_year, new_ytm, frequency, periods)
print(f"5. Actual New Bond Price: {actual_new_bond_price:.2f}")

# Compare the estimated bond price to the actual bond price
price_difference = actual_new_bond_price - estimated_new_bond_price
print(f"Price Difference: {price_difference:.2f}")


In [ ]:
# 缝合bondA,B变成一个new zero coupon bond
#Determine the price of a 5 year, $10,000 zero-coupon bond using the informaion below.
"""
Bond A
- Face value: $10,000
- Maturity: 5 years
- Coupon rate: 6% per annum (paid annually)
- Price:$11,000

Bond B
- Face value: $10,000
- Maturity: 5 years
- Coupon rate: 4% per annum (paid annually)
- Price: $9,500
"""

import numpy as np

# Bond A
face_value_A = 10000
maturity_A = 5
coupon_rate_A = 0.06
price_A=11000
#frequency = 2

# Bond B
face_value_B = 10000
maturity_B = 5
coupon_rate_B = 0.04
price_B=9500


# Create a matrix equation
A = np.array([[10000, 10000], 
            [coupon_rate_A, coupon_rate_B]])
            
b = np.array([10000, 0])

# Solve the matrix equation
weights = np.linalg.solve(A, b)
weight_A, weight_B = weights

print(f"Weight of Bond A: {weight_A:.2f}")
print(f"Weight of Bond B: {weight_B:.2f}")

# Calculate the face value of the zero-coupon bond needed to cancel out the coupon payments
price_zero = (weight_A * price_A + weight_B * price_B)
print(f"Price of a 5 year, zero-coupon bond with a facevalue of $10,000: ${price_zero:.2f}"


## Problem 1:

An investment firm wants to immunize a pension fund's liability. The pension fund has a liability of $10,000,000 due in 10 years. The firm decides to invest in bonds to match the Macaulay duration of the portfolio with the time horizon of the liability. 

The firm considers two bonds:

Bond A:
- Face value: $1,000
- Coupon rate: 4% per annum, paid semi-annually
- Time to maturity: 10 years

Bond B:
- Face value: $1,000
- Coupon rate: 6% per annum, paid semi-annually
- Time to maturity: 30 years

Current market yields for both bonds are 5% per annum, compounded semi-annually.

1. Calculate the Macaulay duration of both bonds.
2. Calculate the Macaulay duration of the liability.
3. Construct a bond portfolio with a Macaulay duration equal to the liability's duration by investing in Bonds A and B. Determine the investment amount in each bond.
4. Calculate the initial investment required for the bond portfolio.
5. Suppose the yield curve shifts by 1% (up or down). Calculate the change in the bond portfolio's value.


In [ ]:
#immunization

import numpy as np

def bond_price(F, C, lamb, m, n):
    '''
    F is the face value of the bond
    C is the yearly coupon payment
    lamb is the yield of the bond 
    m is the number of coupon payments per year
    n is the number of coupon payments remaning
    '''
    discount_factor = (1 + lamb / m) ** n
    bond_price = F / discount_factor + (C / lamb) * (1 - 1 / discount_factor)
    return bond_price

def macaulay_duration(c, y, m, n):
    """
    c is the coupon rate per period
    y is the yield per period
    m is the number of periods per year
    n is the number of periods remaining
    """
    
    D = ((1+y)/(m*y)) - (1+y+n*(c-y))/(m*c*((1+y)**n-1)+m*y)
    return D

# Bond information
face_value = 1000
yield_per_annum = 0.05
coupon_rate_A = 0.04
coupon_rate_B = 0.06
maturity_A = 10
maturity_B = 30
payment_frequency = 2

# Calculate the initial investment required for the bond portfolio
price_A = bond_price(face_value, face_value * coupon_rate_A , yield_per_annum, payment_frequency, maturity_A * payment_frequency)
price_B = bond_price(face_value, face_value * coupon_rate_B , yield_per_annum, payment_frequency, maturity_B * payment_frequency)

# Calculate the Macaulay duration for both bonds
duration_A = macaulay_duration(coupon_rate_A / payment_frequency, yield_per_annum/ payment_frequency, payment_frequency, maturity_A * payment_frequency)
duration_B = macaulay_duration(coupon_rate_B / payment_frequency, yield_per_annum/ payment_frequency, payment_frequency, maturity_B * payment_frequency)

# Calculate the Macaulay duration of the liability
liability = 10_000_000
liability_due_in_years = 10
price_liability=liability*(1/(1+yield_per_annum))**liability_due_in_years
liability_duration = liability_due_in_years  # Since it's like a zero-coupon bond
#如果不是的话要算一下macaulay_duration
#duration_A = macaulay_duration(coupon_rate_A / payment_frequency, yield_per_annum/ payment_frequency, payment_frequency, maturity_A * payment_frequency)

# Construct a bond portfolio
# We will solve the following linear system of equations to find the weights of Bonds A and B:
# weight_A * duration_A + weight_B * duration_B = liability_duration
# weight_A + weight_B = 1
A = np.array([[price_A, price_B], [price_A*duration_A, price_B*duration_B]])
B = np.array([price_liability, price_liability*liability_duration])

# X1 and X2 are the number of Bonds A and B to buy.
X1, X2= np.linalg.solve(A, B)

initial_investment_A=X1*price_A
initial_investment_B=X2*price_B

total_initial_investment = initial_investment_A + initial_investment_B

# Calculate the change in the bond portfolio's value if the yield curve shifts by 1% (up or down)
shift = 0.01
new_yield = yield_per_annum + shift
new_price_A = bond_price(face_value, face_value * coupon_rate_A , new_yield, payment_frequency, maturity_A  * payment_frequency)
new_price_B = bond_price(face_value, face_value * coupon_rate_B , new_yield, payment_frequency, maturity_B  * payment_frequency)
new_price_liability=liability*(1/(1+new_yield))**liability_due_in_years

new_value_A = initial_investment_A * new_price_A / price_A
new_value_B = initial_investment_B * new_price_B / price_B
new_portfolio_value = new_value_A + new_value_B

change_in_value = new_portfolio_value - total_initial_investment

# Print results
print(f"Macaulay Duration of Bond A: {duration_A:.2f} years")
print(f"Macaulay Duration of Bond B: {duration_B:.2f} years")
print(f"Macaulay Duration of Liability: {liability_duration} years")
print(f"Initial investment in Bond A: ${initial_investment_A:,.2f}")
print(f"Initial investment in Bond B: ${initial_investment_B:,.2f}")
print(f"Initial Present Value of Obligation: ${price_liability:,.2f}")
print(f"Total initial investment: ${total_initial_investment:,.2f}")
print(f"New investment in Bond A: ${new_value_A:,.2f}")
print(f"New investment in Bond B: ${new_value_B:,.2f}")
print(f"Total new investment: ${new_portfolio_value:,.2f}")
print(f"New Present Value of Obligation: ${new_price_liability:,.2f}")


In [ ]:
#calculate spot rate and forward rate and predict the next S5
S1 = 0.045
#S2
S3 = 0.062
S4 = 0.068
#S5
#f12
f23 = 0.074
f34 = 0.086
#f45
coupon_rate_X=.05
coupon_rate_Y=.07
price_X=92.23
price_Y=100.89
S2=np.sqrt((1+S3)**3/(1+f23))-1
f12 = (1+S2)**2/(1+S1)-1
#S2,S4->f24
#f24 = (((1+S4)**4)/((1+S2)**2))**(1/(4-2)) -1
#S2,f24->S4
#S4=(((1+S2)**2)*((1+f24)**(4-2)))**(1/4)-1
#S4,f24->S2
#S2=np.sqrt((1+S4)**4/(1+f24)**(4-2))-1
#
weights = np.linalg.solve(A, b)
weight_X, weight_Y = weights
print(weight_X, weight_Y)
# Calculate the face value of the zero-coupon bond needed to cancel out the coupon payments
price_zero = (weight_X * price_X + weight_Y * price_Y)
print(price_zero)
S5=(100/price_zero)**(1/5)-1
f45 = (1+S5)**5/(1+S4)**4-1
print(f"S2: {S2:.4f}")
print(f"f12: {f12:.4f}")
print(f"S5: {S5:.4f}")
print(f"f45: {f45:.4f}")

## Problem 2

Consider an obligation with the following cash flow stream (yearly): $a = [400, 800, 700, 600, 200, 150, 150, 100]$.

We have two bonds available with annual payments to construct an immunized portfolio:

1. Bond 1:  Yearly coupon payment = 5, Face value = 100, Time to maturity = 10 years
2. Bond 2:  Yearly coupon payment = 12, Face value = 100, Time to maturity = 6 years

The yearly spot rates are given as: $s = [7.20, 8.10, 8.70, 9.25, 9.70, 10.10, 10.45, 10.80, 11.10, 11.38, 11.62, 11.83]$ (in percentage).

### Task

Construct an immunized portfolio using the given bonds and obligation cash flow stream. 

1. Calculate the present value of the obligation.
2. Calculate the quasi-modified duration of the obligation.
3. Calculate the quasi-modified duration of Bond 1 and Bond 2.
4. Solve the system of equations to find the appropriate weights for the bonds in the portfolio.

In [ ]:
#immunization and quasi-modified duration of obligation and bond1,2
import numpy as np

# Yearly spot rates
s = np.array([7.20, 8.10, 8.70, 9.25, 9.70, 10.10, 10.45, 10.80, 11.10, 11.38, 11.62, 11.83]) / 100

# Obligation cash flow stream
a = np.array([400, 800, 700, 600, 200, 150, 150, 100])

# Bond information
c1 = 5
c2 = 12
F = 100
T1 = 10
T2 = 6

#cash flows of the bonds
flow1 = np.concatenate([c1 * np.ones(T1 - 1), [F + c1]])
flow2 = np.concatenate([c2 * np.ones(T2 - 1), [F + c2]])

def present_value(cash_flow, spot_rates):
    """Calculate the present value of a cash flow stream given the spot rates."""
    n = len(cash_flow)
    pv = 0
    
    for i in range(n):
        discount_factor = (1 + spot_rates[i]) ** -((i + 1))
        pv += cash_flow[i] * discount_factor
        
    return pv

def quasi_modified_duration(cash_flow, spot_rates):
    """Calculate the quasi-modified duration of a cash flow stream (for m=1) given the spot rates."""
    n = len(cash_flow)
    pv = present_value(cash_flow, spot_rates)
    duration = 0
    
    for i in range(n):
        time_weighted_cash_flow = (i + 1) * cash_flow[i]
        discount_factor = (1 + spot_rates[i]) ** (-(i + 2))
        duration += time_weighted_cash_flow * discount_factor
        
    return duration / pv

# Calculate the present value and quasi-modified duration of the obligation
PV = present_value(a, s)
D = quasi_modified_duration(a, s)


# Calculate the present value and quasi-modified duration of Bond 1
P1=present_value(flow1, s)
D1 = quasi_modified_duration(flow1, s)

# Calculate the present value and quasi-modified duration of Bond 2
P2=present_value(flow2, s)
D2 = quasi_modified_duration(flow2, s)

# Solve the system of equations
A = np.array([[P1, P2], [P1 * D1, P2 * D2]])
B = np.array([PV, PV * D])
X = np.linalg.solve(A, B)

print("Weights for the immunized portfolio:")
print(f"Bond 1: {X[0]:.2f}")
print(f"Bond 2: {X[1]:.2f}")


### dis3(改)Problem 1

A recent graduate has a student loan of $30,000 with an interest rate of 6% per annum, compounded monthly. The loan is to be repaid in equal monthly installments over a 10-year period. After making payments for 2 years, the graduate has an opportunity to refinance the remaining balance of the loan at a reduced interest rate of 4.5% per annum, compounded monthly. The new loan would  have a repayment period of 15 years and be paid in equal monthly installments.

Suppose the graduate has a risk-free savings account that pays 3% per annum, compounded monthly. Should the graduate refinance the student loan? Calculate the difference in monthly payments before and after refinancing, as well as the net present value (NPV) of the savings from refinancing at the original 6% interest rate.

- Original loan amount: $30,000
- Original interest rate: 6% per annum, compounded monthly
- Original repayment period: 10 years

- Time elapsed: 2 years
- Remaining balance: to be calculated

- New interest rate: 4.5% per annum, compounded monthly
- New repayment period: 15 years

- Risk-free savings account interest rate: 3% per annum, compounded monthly

In [ ]:
#换了两年后利率降低了 dis3-1

### dis3 Problem 2

In 2008, at the height of the financial crisis, a homeowner had a variable-rate mortgage on their home. The mortgage value was $500,000, with a term of 25 years and monthly compounding. Initially, the nominal annual interest rate was 3%, which was fixed for the first 3 years. After that period, the interest rate would be adjusted according to the prevailing rates.

Due to the financial crisis, after the 3-year fixed-rate period, the interest rate skyrocketed to 8%. The homeowner was given two options to cope with this change: either increase the monthly payment amount to accommodate the new interest rate or keep the monthly payment amount the same and extend the length of the mortgage.

1. Calculate the original monthly payment amount when the interest rate was 3%.
2. Calculate the new monthly payment amount if the homeowner decides to increase the monthly payment to accommodate the new interest rate of 8%.
3. If the homeowner chooses to keep the monthly payment amount the same, how many months beyond the original term of 25 years does the mortgage need to be extended to pay off the loan? Round your numerical answer to the nearest integer number of months.


In [ ]:
#换了三年后利率升高，要求月付相同增加期限或增加月还款额，注意这里是25年以外还要77年


**dis3-3 10. Variable-Rate Mortgages**  
The Smith family just took out a variable-rate mortgage on their new home. The mortgage value is $100,000, the term is 30 years, and initially the interest rate is 8%. The interest rate is guaranteed for 5 years, after which time the rate will be adjusted according to prevailing rates. The new rate can be applied to their loan either by changing the payment amount or by changing the length of the mortgage.

(a) What is the original monthly mortgage payment? (Assume payments are yearly.)  
(b) What will be the mortgage balance after 5 years?  
(c) If the interest rate on the mortgage changes to 9% after 5 years, what will be the new yearly payment that keeps the termination time the same?  
(d) Under the interest change in (c), what will be the new term if the payments remain the same?

In [ ]:
#一样，换了5年后利率升高，要求增加月还款额

## dis4-2 Problem 2

Consider a bond with a face value of $1,000, a coupon rate of 6% per annum, and 8 years remaining until maturity. The bond pays interest semi-annually. The current yield to maturity (YTM) is 4% per annum.

1. Calculate the bond price.
2. Calculate the Macaulay duration of the bond.
3. Calculate the modified duration of the bond.
4. If the YTM increases by 1% (i.e., goes from 4% to 5% per annum), estimate the new bond price using the modified duration.
5. Calculate the actual new bond price using the updated YTM, and compare it to the estimated bond price from step 4.

## dis7-1 Problem 1

Imagine you are a student trying to plan your course schedule for the upcoming semester. You have a list of courses you want to take, each with a specific number of credits and an estimated difficulty level. Due to your other commitments, you have a maximum number of credits you can enroll in and a maximum total difficulty level you can handle. Your goal is to maximize your learning outcome, represented as a weighted sum of credits and difficulty for the courses you take.

Here's the list of courses, their credit hours, difficulty levels, and learning outcomes:

| Course      | Credits | Difficulty | Learning Outcome |
|-------------|---------|------------|------------------|
| Math        | 4       | 8          | 20               |
| Physics     | 3       | 7          | 15               |
| Chemistry   | 4       | 6          | 18               |
| Biology     | 3       | 5          | 12               |
| Programming | 5       | 4          | 22               |
| History     | 3       | 3          | 10               |
| Music       | 2       | 2          | 8                |

You can only enroll in a maximum of 12 credit hours, and your maximum total difficulty level is 20. Which courses should you enroll in to maximize your learning outcome?

Formulate this problem as an integer linear programming problem and solve it using Python.

**Note:** The decision variables in this problem are binary, with a value of 1 indicating that you enroll in a course and 0 indicating that you do not enroll in the course. The objective function is the sum of the learning outcomes of the enrolled courses, and the constraints are the maximum credit hours and difficulty levels.

## dis7-2 Problem 2

Imagine you have a series of future monetary obligations, in dollars, that need to be paid at the end of each year over the next 8 years as follows:

| Year     | 1   | 2   | 3   | 4   | 5   | 6   | 7   | 8   |
|----------|-----|-----|-----|-----|-----|-----|-----|-----|
| Obligation | 150 | 250 | 400 | 550 | 700 | 850 | 1000 | 1200 |

We want to invest now to meet these obligations as they occur. To do this, we must purchase bonds of various maturities and use the coupon payments and redemption values to fulfill the obligations.

Suppose there are 8 bonds available for purchase, each with a face value of $ 100  and a price equal to $100. Each of the 8 bonds is an annual coupon bond with its annual coupon in percentage equal to its maturity in years multiplied by 0.5. The maturities span 1 to 8 years, and the annual coupons range from 0.5% to 4%.

To be clear, we have available a 1-year 0.5% annual coupon bond, a 2-year 1% annual coupon bond, a 3-year 1.5% annual coupon bond, all the way up to an 8-year 4% annual coupon bond, for a total of 8 bonds.

The objective function to be minimized is the total cost of the portfolio, which is equal to the sum of the prices of the bonds times the number of units purchased.

**Question:** How many units of the 8-year 4% annual coupon bond should be purchased in the optimal portfolio? Round your numerical answer to the nearest integer number of units.


## dis 7-3 Problem 3
A portfolio manager wants to allocate funds to a set of 4 assets in order to maximize the expected return of the portfolio while limiting the risk exposure. The expected annual returns for the 4 assets are as follows:

| Asset | Expected Annual Return |
|-------|------------------------|
| A     | 10%                    |
| B     | 15%                    |
| C     | 12%                    |
| D     | 8%                     |

The portfolio manager wants to ensure that at least 20% of the total investment is allocated to each asset to maintain diversification.

Additionally, to limit risk exposure, the total investment in assets B and C should not exceed 60% of the total investment.

The manager has a budget of \$100,000 to invest in the assets.

Formulate a linear programming problem to determine the optimal allocation of funds among the 4 assets to maximize the expected return of the portfolio while satisfying the given constraints.

## *dis7-4 Problem 4
A company has identified several promising projects as indicated in the table below. The cash flows for the first 2 years are shown (they are all negative). The cash flows in later years are positive, and the net present value (NPV) of each project is shown. The company managers have decided that they can allocate up to 300,000 in each of the first 2 years to fund these projects. If less than 300,000 is used in the first year, the balance can be invested at 12% and used to augment the next year's budget. Which projects should be funded?

| Project | Cash flow (Year 1) | Cash flow (Year 2) | NPV |
|---------|--------------------|--------------------|-----|
| 1       | -\$100,000         | -\$50,000          | \$200,000 |
| 2       | -\$150,000         | -\$50,000          | \$250,000 |
| 3       | -\$200,000         | -\$100,000         | \$500,000 |
| 4       | -\$50,000          | -\$50,000          | \$120,000 |
| 5       | -\$100,000         | -\$200,000         | \$450,000 |


In [ ]:
#dis8不想看额，还有midterm